### PR TITLE
TS conversion for `LoginViewModel`

### PR DESCRIPTION
**Pull request by ibeckermayer**

### CODE REVIEW

File: .eslintrc.js
Comment (ibeckermayer) [inline .eslintrc.js]: I noticed this was the de facto standard so I made it de jure.

File: src/domain/ViewModel.ts
Comment (ibeckermayer) [inline src/domain/ViewModel.ts]: I made this change in order to get [this line](https://github.com/vector-im/hydrogen-web) to compile.

**Comment by ibeckermayer** [comment]
Also of note:

Unfortunately when a variable is typed as a Javascript class like [`private _passwordLoginViewModel?: PasswordLoginViewModel;`](https://github.com/vector-im/hydrogen-web/pull/777/files#diff-fdb2dad635bedc87b24b90bdff4fa083c7f69bd7431fe60f0c0e4d1ff92de1d0R37), the typescript compiler isn't well equipped to deal with it and falls back to an `any`:
```
Could not find a declaration file for module './PasswordLoginViewModel.js'. './src/domain/login/PasswordLoginViewModel.js' implicitly has an 'any' type.ts(7016)
```

In the case of these optional types (`?:`), this means that we can use them in a way that would normally cause the compiler to complain about us not checking for `undefined`. Hypothetically I could manually identify and solve for all these now, however I decided that because this code is already battle tested, we can safely ignore this issue for the moment. Once we get around to converting `PasswordLoginViewModel.js` to `PasswordLoginViewModel.ts`, the compiler will begin complaining and we can restructure things then.

File: src/domain/ViewModel.ts
Comment (bwindels) [inline src/domain/ViewModel.ts]: can you remove this commit together with the one that introduced the change (and force push)?

File: src/domain/ViewModel.ts
Comment (bwindels) [reply src/domain/ViewModel.ts]: which line? the link just points to the repo

File: src/domain/login/LoginViewModel.ts
Comment (bwindels) [inline src/domain/login/LoginViewModel.ts]: Please don't use default exports anywhere, see https://github.com/vector-im/hydrogen-web/blob/master/codestyle.md

File: src/domain/login/LoginViewModel.ts
Comment (bwindels) [inline src/domain/login/LoginViewModel.ts]: These can all be `import type` I imagine

File: src/domain/login/LoginViewModel.ts
Comment (bwindels) [inline src/domain/login/LoginViewModel.ts]: OOI, why do you do this? Because it's an async method?

File: src/domain/login/LoginViewModel.ts
Comment (bwindels) [inline src/domain/login/LoginViewModel.ts]: Should this be `ILoginMethod` rather?

File: src/domain/login/LoginViewModel.ts
Comment (bwindels) [inline src/domain/login/LoginViewModel.ts:197]: good catch, we were leaking `_startSSOLoginViewModel` so far then?

File: src/domain/login/LoginViewModel.ts
Comment (bwindels) [inline src/domain/login/LoginViewModel.ts]: can you add return types to all the getters and methods as well please?

File: src/domain/login/LoginViewModel.ts
Comment (bwindels) [inline src/domain/login/LoginViewModel.ts:40]: usually we add the type annotation for props anyway, regardless of whether typescript can infer it or not, as it's a bit easier to parse for the humans ...

**Review by bwindels — COMMENTED** [review]
This looks great, thank you for your PR. Just a few nits about our (not always documented) habits.

File: src/domain/ViewModel.ts
Comment (ibeckermayer) [reply src/domain/ViewModel.ts]: Oops. Meant to link [here](https://github.com/ibeckermayer/hydrogen-web/blob/5c5c6007047e10cf0dc00c9d9c5be1b77db99eae/src/domain/login/LoginViewModel.ts#L182).

File: src/domain/login/LoginViewModel.ts
Comment (ibeckermayer) [reply src/domain/login/LoginViewModel.ts]: Right, that was more of a note to self and not using proper terminology (although I used it in the commit message, which isn't ideal). What I meant was add them to `index.ts` so that I didn't need to import from each file separately, see

https://github.com/vector-im/hydrogen-web/pull/777/commits/5c5c6007047e10cf0dc00c9d9c5be1b77db99eae

File: src/domain/login/LoginViewModel.ts
Comment (ibeckermayer) [reply src/domain/login/LoginViewModel.ts]: Yes, I've been using the `.ts-eslintrc.js` which was complaining at me
```
Promises must be handled appropriately or explicitly marked as ignored with the `void` operator.eslint[@typescript-eslint/no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/v4.29.2/packages/eslint-plugin/docs/rules/no-floating-promises.md)
```

Normally I would mark the function `async` and `await` it, but it's a constructor so I just `void`ed it.

File: src/domain/login/LoginViewModel.ts
Comment (ibeckermayer) [reply src/domain/login/LoginViewModel.ts:197]: It seems so, you were effectively calling
```javascript
        this._startSSOLoginViewModel = this.disposeTracked(undefined);
```

File: src/domain/login/LoginViewModel.ts
Comment (ibeckermayer) [reply src/domain/login/LoginViewModel.ts]: Ah yes, makes sense. Double check me here -- I think [this](https://github.com/ibeckermayer/hydrogen-web/blob/f8c47059447cbc12962f12a6373b49cf634cb721/src/domain/login/LoginViewModel.ts#L112) is meant to be `TokenLoginMethod` and then [this](https://github.com/ibeckermayer/hydrogen-web/blob/f8c47059447cbc12962f12a6373b49cf634cb721/src/domain/login/LoginViewModel.ts#L126) is `PasswordLoginMethod`.

File: src/domain/RootViewModel.js
Comment (MidhunSureshR) [inline src/domain/RootViewModel.js]:
```suggestion
import {LoginViewModel} from "./login/LoginViewModel";
```

File: src/domain/login/LoginViewModel.ts
Comment (MidhunSureshR) [reply src/domain/login/LoginViewModel.ts]: That is correct :+1:

File: src/domain/login/LoginViewModel.ts
Comment (MidhunSureshR) [inline src/domain/login/LoginViewModel.ts]: Why `await` here and in other places? The intent is to not block further execution unless there's a reason to.

File: src/domain/login/LoginViewModel.ts
Comment (MidhunSureshR) [inline src/domain/login/LoginViewModel.ts]: Can you add a newline here?

File: src/domain/ViewModel.ts
Comment (MidhunSureshR) [reply src/domain/ViewModel.ts]: The argument passed to the `emitChange` method does nothing as of now but it will serve some purpose if Hydrogen is ported to another platform. So we'd like to have typescript shout if we forget to pass an argument to the method.

For [the call here](https://github.com/ibeckermayer/hydrogen-web/blob/5c5c6007047e10cf0dc00c9d9c5be1b77db99eae/src/domain/login/LoginViewModel.ts#L182), you can do `emitChange("loginViewModels")`.

**Review by MidhunSureshR — COMMENTED** [review]
Thank you, leaving some comments here.

You'll also need to [sign-off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off) before we can merge this PR.

File: src/domain/login/LoginViewModel.ts
Comment (ibeckermayer) [reply src/domain/login/LoginViewModel.ts]: I wasn't thinking clearly, removed these in 40276298041a70058f5447795d591f48af83ac96

**Comment by ibeckermayer** [comment]
https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off

Signed-off-by: Isaiah Becker-Mayer isaiah@becker-mayer.com

File: src/domain/login/LoginViewModel.ts
Comment (MidhunSureshR) [inline src/domain/login/LoginViewModel.ts]: Can you mark methods starting with _ as private?

**Review by MidhunSureshR — APPROVED** [review]
Just one more thing:

**Comment by MidhunSureshR** [comment]
Looks like the linter is picking up lots of errors due to the new rules. Might make sense to add these rules in a separate PR which also fixes all these errors.

**Comment by ibeckermayer** [comment]
@MidhunSureshR added those `private`s and reverted all the eslint changes for now. Looks ready to merge to me (you will have to do so since I don't have write access to the repo).

**Pull request merged**
